src/gstring.rs

```diff
@@ -0,0 +1,333 @@
+    unsafe fn from_glib_full(ptr: *mut i8) -> Self {
+        GString::new(ptr as *const c_char)
+    }
+}
+
+impl FromGlibPtrNone<*const c_char> for GString {
+    #[inline]
+    unsafe fn from_glib_none(ptr: *const c_char) -> Self {
+        let cstr = CStr::from_ptr(ptr);
+        GString::from_cstring(cstr.to_owned())
+    }
+}
+
+impl FromGlibPtrNone<*mut u8> for GString {
+    #[inline]
+    unsafe fn from_glib_none(ptr: *mut u8) -> Self {
+        let cstr = CStr::from_ptr(ptr as *mut i8);
+        GString::from_cstring(cstr.to_owned())
+    }
+}
+
+impl FromGlibPtrNone<*mut i8> for GString {
+    #[inline]
+    unsafe fn from_glib_none(ptr: *mut i8) -> Self {
+        let cstr = CStr::from_ptr(ptr);
+        GString::from_cstring(cstr.to_owned())
+    }
+}
+
+impl<'a> ToGlibPtr<'a, *const c_char> for GString {
+    type Storage = Option<CString>;
+
+    #[inline]
+    fn to_glib_none(&'a self) -> Stash<'a, *const c_char, Self> {
+        let cstring = self.to_cstring();
+        Stash(cstring.as_ptr(), Some(cstring))
+    }
+
+    #[inline]
+    fn to_glib_full(&self) -> *const c_char {
+        unsafe {
+            glib_ffi::g_strndup(self.as_ptr() as *const c_char, self.len() as libc::size_t)
+                as *const c_char
+        }
+    }
+}
+
+impl<'a> ToGlibPtr<'a, *mut c_char> for GString {
+    type Storage = Option<CString>;
+
+    #[inline]
+    fn to_glib_none(&'a self) -> Stash<'a, *mut c_char, Self> {
+        let cstring = self.to_cstring();
+        Stash(cstring.as_ptr() as *mut c_char, Some(cstring))
+    }
+
+    #[inline]
+    fn to_glib_full(&self) -> *mut c_char {
+        unsafe {
+            glib_ffi::g_strndup(self.as_ptr() as *const c_char, self.len() as libc::size_t)
+                as *mut c_char
+        }
+    }
+}
+
+impl GlibPtrDefault for GString {
+    type GlibType = *const c_char;
+}
+
+impl StaticType for GString {
+    fn static_type() -> Type {
+        std::string::String::static_type()
+    }
+}
+
+impl StaticType for Vec<GString> {
+    fn static_type() -> Type {
+        unsafe {
+            from_glib(glib_ffi::g_strv_get_type())
+        }
+    }
+}
+
+impl<'a> FromValueOptional<'a> for GString {
+    unsafe fn from_value_optional(value: &'a Value) -> Option<Self> {
+        let val = value.to_glib_none().0;
+        if val == ptr::null() {
+            None
+        } else {
+            let ptr = gobject_ffi::g_value_dup_string(val);
+            Some(GString::new(ptr))
+        }
+    }
+}
+
+impl_from_glib_container_as_vec_string!(GString, *const c_char);
+impl_from_glib_container_as_vec_string!(GString, *mut c_char);
+
+#[cfg(test)]
+mod tests {
+    use gstring::GString;
+    use std::ffi::CString;
+
+    #[test]
+    fn test_holder() {
+        let data = CString::new("foo").unwrap();
+        let ptr = data.into_raw();
+
+        unsafe {
+            let holder = GString::new(ptr);
+            assert_eq!(holder.as_str(), "foo");
+            let foo: Box<str> = holder.into();
```

Comment:
If `impl From<GString> for Box<str>` called on "owned ptr" case, then seems it needed `mem::forget` too,
maybe `impl From<GString> for String` too, but not sure

Reply:
I'm sorry, I'm not sure to understand what you mean, can you please clarify?

Reply:
Missed `let ptr_copy = glib_ffi::g_strdup(ptr);` in this test